src/main/java/com/sarality/db/SQLiteDatabaseProvider.java

```diff
@@ -63,5 +67,98 @@ public void onCreate(SQLiteDatabase database) {
   @Override
   public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
     logger.info("Upgrading tables for database {} to version {}.", dbName, dbVersion);
+
+    for (TableDefinition definition : definitionList) {
+      // get current schema
+      List<Column> versionColumns = new ArrayList<>();
+      boolean isSchemaInitialized = false;
+
+      for (Integer version = oldVersion + 1; version <= newVersion; version++) {
+        Map<UpdateOperationType, List<UpdateOperation>> operations = definition.getUpdatesForVersion(version);
+        if (operations == null) {
+          logger.info("No version upgrade operations for table {} in version {}.", definition.getTableName(), version);
+          continue;
+        }
+
+        logger.info("Running upgrade operations for table {} in version {}.", definition.getTableName(), version);
+
+        if (!isSchemaInitialized) {
+          versionColumns = getSchemaForVersion(oldVersion, definition);
+          isSchemaInitialized = true;
+        }
+
+        // is this a new table in this version?
+        List<UpdateOperation> updatesList = operations.get(UpdateOperationType.CREATE_TABLE);
+
+        if (updatesList != null && updatesList.size() > 0) {
+          // Create the table - sql
+          logger.info("Create Table SQL: {} ", TableSQLGenerator.getCreateSql(definition.getTableName(),
+              definition.getColumns()));
+
+          db.execSQL(TableSQLGenerator.getCreateSql(definition.getTableName(),
+              definition.getColumns()));
+
+          // I can safely ignore all subsequent updates now
+          break;
+        }
+
+        //are there any new columns added?
+        updatesList = operations.get(UpdateOperationType.ADD_COLUMN);
+        if (updatesList != null && updatesList.size() > 0) {
+
+          for (UpdateOperation updateOperation : updatesList) {
+            versionColumns.add(updateOperation.getNewColumn());
+            logger.info("Add Column SQL: {} ",
+                TableSQLGenerator.getAddColumnSql(definition.getTableName(),
+                    updateOperation.getNewColumn()));
+
+            db.execSQL(
+                TableSQLGenerator.getAddColumnSql(definition.getTableName(),
+                    updateOperation.getNewColumn()));
+
+          }
+        }
+      }
+    }
   }
+
+
+  private List<Column> getSchemaForVersion(int oldVersion, TableDefinition definition) {
```

Comment:
that actually isnt how it was working, but as discussed, makes sense to have a versionSchemaProvider that catalogs the schemas by version and can provide it to the table updater as required. Added the class and modified this method.